RavenDB-24327 Persist freed pages in journal files  to skip their checksum validation after restart.

arekpalinski opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-24327/Invalid-checksum-error-after-process-termination

### Additional description

Added support for tracking freed pages in transactions and persisting this information in the journal. This ensures proper recovery handling by skipping checksum checks on freed pages, in particular those that were part of a sparse region.

### Type of change

- [x] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [ ] New feature

### How risky is the change?

- [x] Low 
- [ ] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [ ] Non breaking change
- [x] Ensured. Please explain how has it been implemented?
   - Journal entries that include free pages set a new flag in a transaction header
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing
- [ ] Existing tests verify the correct behavior

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

arekpalinski committed:
RavenDB-24327 Persist freed pages in journal files  to skip their checksum validation after restart.

Added support for tracking freed pages in transactions and persisting this information in the journal. This ensures proper recovery handling by skipping checksum checks on freed pages, in particular those that were part of a sparse region.

arekpalinski commented:
@ravendb-bot test this in debug please

arekpalinski commented:
Please hold on with the review

arekpalinski commented:
Ready for the review

ayende commented on src/Voron/Impl/Journal/WriteAheadJournal.cs:
```diff
@@ -2084,8 +2057,31 @@ private Pal.journal_entry PrepareToWriteToJournal(LowLevelTransaction tx, ref Pa
+                        {
+                            freedPages.Remove(page.PageNumberInDataFile + i);
+                        }
+                    }
```
Shouldn't this be something that we are handling during the `ModifyPage`, instead?

redknightlois replied:
Pushing this into the transaction would require us to keep the state in sync during a high traffic operation (`AllocatePageImpl`) while we can do it here with `O(#tx pages + size of freed set)`. I don't know, from the performance point of view it could make sense to push it earlier since we have `O( 1 )` on average by accessing the `HashSet` anyways but definitely something that should be shown with performance numbers in hand. I see potential cases where it would be interesting to see like on batch deletes.

ayende replied:
The problem is that it is entirely feasible to have a transaction in which we:

* modify the page
* free it
* allocate it again
* free it again

Etc...

That is a problem for this scenario, because you cannot assume that modified page / free pages are exclusive.

redknightlois replied:
We need a test for that. That's a very interesting edge case I was not thinking about.

arekpalinski replied:
After direct discussion with @ayende we agreed to handle it this way. It's way simpler approach than syncing transaction pages and free ones during transaction. Also the same approach we had in versions before 8.0.

The code itself got fixed and the test case was added.

redknightlois commented on src/Voron/Impl/Journal/TransactionHeader.cs:
```diff
@@ -92,4 +92,11 @@ public override string ToString()
-}
\ No newline at end of file
+    
+    [StructLayout(LayoutKind.Explicit, Pack = 1)]
```
`LayoutKind.Explicit` makes `Pack` irrelevant. In this case, `Pack` is a no-op can be safely removed.

ayende replied:
Note, we should probably `pack` this using `pfor` or the like.

arekpalinski replied:
Removed `Pack`. The code got changed to use PFor encoding.

redknightlois commented on src/Voron/Impl/Journal/WriteAheadJournal.cs:
```diff
@@ -2084,8 +2057,31 @@ private Pal.journal_entry PrepareToWriteToJournal(LowLevelTransaction tx, ref Pa
+                
+                foreach (var page in txPages)
+                {
+                    if (freedPages.Remove(page.PageNumberInDataFile) && page.NumberOfPages > 1)
```
Invert this, we want to use the short-circuit to avoid the HashSet operation when skipping the loop completely is possible via `page.NumberOfPages`

arekpalinski replied:
This got removed

ayende commented on src/Voron/Impl/Journal/WriteAheadJournal.cs:
```diff
@@ -2084,8 +2057,31 @@ private Pal.journal_entry PrepareToWriteToJournal(LowLevelTransaction tx, ref Pa
+            
+            var freedPagesOverhead = numberOfFreedPages * sizeof(TransactionFreePage);
+            var freedPagesOverheadInPages = checked(freedPagesOverhead / Constants.Storage.PageSize + (freedPagesOverhead % Constants.Storage.PageSize == 0 ? 0 : 1));
+            
```
freedPagesOverhead = encoder.Encode(l, array.Length);

ayende replied:
(int count, int sizeUsed) = encoder.Write(buffer, Container.MaxSizeInsideContainerPage);

arekpalinski committed:
RavenDB-24327 Handle large numbers of freed pages exceeding single PFor encoder write limit

arekpalinski commented:
@ravendb-bot test this in debug please

redknightlois reviewed (approved):
I like that we just needed to impact `FreePage` with an O( 1 ) on average operation. LGTM

ayende commented on src/Voron/Impl/Journal/JournalReader.cs:
```diff
@@ -237,11 +243,76 @@ public bool ReadOneTransactionToDataFile(ref Pager.State dataPagerState, ref Pag
+
+            var header = *(FreePagesHeader*)headerPtr;
+            var sectionHeaders = (EncodedFreePagesSection*)(headerPtr + sizeof(FreePagesHeader) + header.EncodedSectionsSize);
+
```
Let's add some checks here that we aren't trying to read past the transaction size.

arekpalinski replied:
Added

ayende commented on src/Voron/Impl/Journal/JournalReader.cs:
```diff
@@ -237,11 +243,76 @@ public bool ReadOneTransactionToDataFile(ref Pager.State dataPagerState, ref Pag
+            {
+                var sectionSize = sectionHeaders[i].Size;
+                
+                using var reader = new FastPForBufferedReader(allocator);
```
Does this free all the memory?
Do we need to *reset* the allocator between journals?

arekpalinski replied:
Yes, it does. Anyway the allocator is reset now between journals

ayende commented on src/Voron/Impl/Journal/WriteAheadJournal.cs:
```diff
@@ -353,7 +354,7 @@ public bool RecoverDatabase(TransactionHeader* txHeader, out long lastJournalNum
                     List<TransactionHeader> transactionHeaders;
                     var journalReader = new JournalReader(_env, journalNumber, journalPager, journalPagerState, dataPager, recoveryPager, modifiedPages, logInfo, currentFileHeader,
-                        transactionHeader);
+                        transactionHeader, allocator);
```
can we reset here to avoid leaks?

arekpalinski replied:
Done

ayende commented on src/Voron/Impl/Journal/WriteAheadJournal.cs:
```diff
@@ -2268,6 +2273,9 @@ private Pal.journal_entry PrepareToWriteToJournal(LowLevelTransaction tx, ref Pa
             txHeader.JournalId = _headerAccessor.JournalId;
             
+            if (freedPages is { Count: > 0 })
+                txHeader.Flags |= TransactionPersistenceModeFlags.HasFreePages;
```
Can we move this to line `2180`, where we also have this?

arekpalinski replied:
Done

ayende commented on src/Voron/Impl/Journal/WriteAheadJournal.cs:
```diff
@@ -2491,5 +2499,66 @@ internal sealed class TestingStuff
+        
+        internal static long EncodeFreePages(HashSet<long> freedPages, byte* dst, ByteStringContext allocator)
+        {
+            var pages = freedPages.ToArray();
```
we already have an allocator, let's use that instead of allocating managed array.
I would like to avoid this allocation each time

arekpalinski replied:
Done

arekpalinski committed:
RavenDB-24327 Enhancements to persisting and reading freed pages in journal files
- used native memory instead of managed ToArray() allocation when writing free pages
- added validation for read bounds
- resetting allocator between journal files during recovery

ayende commented on src/Voron/Impl/Journal/JournalReader.cs:
```diff
@@ -245,7 +245,20 @@ public bool ReadOneTransactionToDataFile(ref Pager.State dataPagerState, ref Pag
+                        $"the transaction uncompressed size is {current->UncompressedSize} bytes");
+                }
+
+                var freePages = ReadEncodedFreePage(freePagesPtr, _allocator);
```
I think we need to pass `UncompressedSize` here as well.
If we have a bad value in one of the segments size, we need to error there as well

arekpalinski replied:
Validation of section sizes added. I put entire validation code inside `ReadEncodedFreePage`

arekpalinski committed:
RavenDB-24327 Added EncodedFreePagesSection size validation. Encapsulate entire reading validation in ReadEncodedFreePage method.

redknightlois commented on src/Voron/Impl/Journal/WriteAheadJournal.cs:
```diff
@@ -2491,5 +2501,72 @@ internal sealed class TestingStuff
         private void RejectCommitsToMerge() => SharedJournalState.SetCancel();
+
+        internal static long EncodeFreePages(HashSet<long> freedPages, byte* dst, ByteStringContext allocator)
+        {
```
Can we add somewhere a bit of information about the format for the journal frame? I am thinking in terms of ourselves 4 years from now.

arekpalinski replied:
I will add a comment showing the format of a journal entry

arekpalinski committed:
RavenDB-24327 Added comment explaining format of a single journal entry

ayende commented on src/Voron/Impl/Journal/WriteAheadJournal.cs:
```diff
@@ -2491,5 +2521,72 @@ internal sealed class TestingStuff
+            {
+                encoder.Encode(pages + processedCount, remainingCount);
+
+                var (count, size) = encoder.Write(output, maxEncodingSectionSize);
```
One last thing - here, can we verify that we have sufficient space?
I _think_ we must have, because we allocated twice as much, but it should be checked.
Also, we may want to error if `count` is 0, etc.

arekpalinski replied:
Done. Please check https://github.com/ravendb/ravendb/pull/21900/commits/b74cf814714454aaec381df25bac63849d35619d

arekpalinski committed:
RavenDB-24327 Validate space availability when encoding free pages in journal. Adjust `EncodeFreePages` method to account for destination buffer size constraints. Prevent overflows and add explicit error handling during encoding.
